cmd/routesrv/main.go

```diff
@@ -0,0 +1,262 @@
+			span.LogKV(
+				"event", "error",
+				"message", msg,
+			)
+		case routesLen > 0:
+			size, initialized = p.b.formatAndSet(routes)
+			if initialized {
+				log.Info("routes initialized")
+				span.SetTag("initialized", true)
+				p.metrics.initializedTimestamp.SetToCurrentTime()
+			} else {
+				log.Debug("routes updated")
+			}
+			p.metrics.updatedTimestamp.SetToCurrentTime()
+			span.SetTag("routes.received_len", routesLen)
+			span.SetTag("routes.stored_bytes", size)
+		}
+
+		span.Finish()
+
+		select {
+		case <-p.quit:
+			return
+		case <-time.After(p.timeout):
+		}
+	}
+}
+
+func newServer(address string, b *eskipBytes, s *eskipBytesStatus) *http.Server {
+	handler := http.NewServeMux()
+
+	handler.Handle("/health", s)
+	handler.Handle("/routes", b)
+	handler.Handle("/metrics", promhttp.Handler())
+
+	return &http.Server{Addr: address, Handler: handler}
+}
+
+func run(o options.Options) error {
+	tracer, err := tracing.InitTracer(o.OpenTracing)
+	if err != nil {
+		return err
+	}
+
+	b := &eskipBytes{tracer: tracer}
+	s := &eskipBytesStatus{b: b}
+
+	dataclient, err := kubernetes.New(kubernetes.Options{
+		KubernetesInCluster:               o.KubernetesInCluster,
+		KubernetesURL:                     o.KubernetesURL,
+		ProvideHealthcheck:                o.KubernetesHealthcheck,
+		ProvideHTTPSRedirect:              o.KubernetesHTTPSRedirect,
+		HTTPSRedirectCode:                 o.KubernetesHTTPSRedirectCode,
+		IngressClass:                      o.KubernetesIngressClass,
+		RouteGroupClass:                   o.KubernetesRouteGroupClass,
+		ReverseSourcePredicate:            o.ReverseSourcePredicate,
+		WhitelistedHealthCheckCIDR:        o.WhitelistedHealthCheckCIDR,
+		PathMode:                          o.KubernetesPathMode,
+		KubernetesNamespace:               o.KubernetesNamespace,
+		KubernetesEnableEastWest:          o.KubernetesEnableEastWest,
+		KubernetesEastWestDomain:          o.KubernetesEastWestDomain,
+		KubernetesEastWestRangeDomains:    o.KubernetesEastWestRangeDomains,
+		KubernetesEastWestRangePredicates: o.KubernetesEastWestRangePredicates,
+		DefaultFiltersDir:                 o.DefaultFiltersDir,
+		OriginMarker:                      o.EnableRouteCreationMetrics,
+		BackendNameTracingTag:             o.OpenTracingBackendNameTag,
+		OnlyAllowedExternalNames:          o.KubernetesOnlyAllowedExternalNames,
+		AllowedExternalNames:              o.KubernetesAllowedExternalNames,
+	})
+	if err != nil {
+		return err
+	}
+	poller := &poller{
+		client:  dataclient,
+		timeout: o.SourcePollTimeout,
+		b:       b,
+		quit:    make(chan struct{}, 1),
+		tracer:  tracer,
+		metrics: newPollerMetrics(),
+	}
+	go poller.poll()
+
+	server := newServer(o.Address, b, s)
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		log.Info("shutting down")
+		close(poller.quit)
+		if err := server.Shutdown(context.Background()); err != nil {
```

Comment:
Do we need shutdown timeout here?

Also from https://pkg.go.dev/net/http#Server.Shutdown

> When Shutdown is called, Serve, ListenAndServe, and ListenAndServeTLS immediately return ErrServerClosed. Make sure the program doesn't exit and waits instead for Shutdown to return.

Reply:
For teardown, if we do the same as in skipper, it's fine. check https://github.com/zalando/skipper/blob/master/skipper.go#L1068-L1080

Shutdown() is safe to use if you are already faded out of the load balancer, so no new connections should be possible if you call Shutdown().
Because we get SIGTERM immediately we should add a gracePeriod, such that our pod is faded out of the endpoints list, that is part of the service on all nodes (iptables change on all nodes where skipper-ingress is running). So the current one is not safe, we need to do a sleep before Shutdown() is called.

Reply:
Got it. I do not really wait for the shutdown to happen properly and this shutdown (if properly awaited) can take a while.